src/visip/dev/dtype_new.py

```diff
@@ -0,0 +1,582 @@
+
+        # place TypeVar argument to the end
+        if 0 <= tv_ind < len(self.args) - 1:
+            a = self.args.pop(tv_ind)
+            self.args.append(a)
+
+
+class Const:
+    def __init__(self, arg):
+        if isinstance(arg, Const):
+            self.arg = arg.arg
+        else:
+            self.arg = arg
+        assert not isinstance(self.arg, Const)
+
+
+class TypeVar:
+    def __init__(self, name):
+        self.name = name
+
+
+class NewType:
+    def __init__(self, name, supertype):
+        self.name = name
+        self.supertype = supertype
+
+
+class Any(metaclass=Singleton):
+    pass
+
+
+class NoneType(metaclass=Singleton):
+    pass
+
+
+def from_typing(type):
+    t, _, _ = from_typing_map(type, {}, {})
+    return t
+
+
+def from_typing_map(type, var_map, new_map):
+    # base
+    if type is int:
+        return Int(), var_map, new_map
+
+    if type is float:
+        return Float(), var_map, new_map
+
+    if type is bool:
+        return Bool(), var_map, new_map
+
+    if type is str:
+        return Str(), var_map, new_map
+
+
+    # TypeVar
+    if typing_inspect.is_typevar(type):
+        if type in var_map:
+            return var_map[type], var_map, new_map
+        t = TypeVar(type.__name__)
+        var_map = var_map.copy()
+        var_map[type] = t
+        return t, var_map, new_map
+
+
+    # NewType
+    if typing_inspect.is_new_type(type):
+        if type in new_map:
+            return new_map[type], var_map, new_map
+        supertype, var_map, new_map = from_typing_map(type.__supertype__, var_map, new_map)
+        t = NewType(type.__name__, supertype)
+        new_map = new_map.copy()
+        new_map[type] = t
+        return t, var_map, new_map
+
+
+    # Tuple
+    if typing_inspect.is_tuple_type(type):
+        args = []
+        for a in typing_inspect.get_args(type, evaluate=True):
+            t, var_map, new_map = from_typing_map(a, var_map, new_map)
+            args.append(t)
+        return Tuple(args), var_map, new_map
+
+
+    # Union
+    if typing_inspect.is_union_type(type):
+        args = []
+        for a in typing_inspect.get_args(type, evaluate=True):
+            t, var_map, new_map = from_typing_map(a, var_map, new_map)
+            args.append(t)
+        return Union(args), var_map, new_map
+
+
+    origin = typing_inspect.get_origin(type)
+
+    # List
+    if origin in [list, typing.List]:
+        t, var_map, new_map = from_typing_map(typing_inspect.get_args(type, evaluate=True)[0], var_map, new_map)
+        return List(t), var_map, new_map
+
+    # Dict
+    if origin in [dict, typing.Dict]:
+        args = typing_inspect.get_args(type, evaluate=True)
+        k, var_map, new_map = from_typing_map(args[0], var_map, new_map)
+        v, var_map, new_map = from_typing_map(args[1], var_map, new_map)
+        return Dict(k, v), var_map, new_map
+
+    # Const
+    if origin is dtype.Constant:
```

Comment:
Just unify Constant and Const, could be named Const.

Reply:
Const[X] - nejprve prevede X na dtype (docasne reseni)

Reply:
Hotovo.